<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="400" alt="Skills Network Logo"  />
    </a>
</p>


# **Data Wrangling Lab**


Estimated time needed: **45 to 60** minutes


In this assignment you will be performing data wrangling.


## Objectives


In this lab you will perform the following:


-   Identify duplicate values in the dataset.

-   Remove duplicate values from the dataset.

-   Identify missing values in the dataset.

-   Impute the missing values in the dataset.

-   Normalize data in the dataset.


<hr>


## Hands on Lab


Import pandas module.


In [2]:
import pandas as pd

<ipython-input-2-7dd3504c366f>:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Load the dataset into a dataframe.


<h2>Read Data</h2>
<p>
We utilize the <code>pandas.read_csv()</code> function for reading CSV files. However, in this version of the lab, which operates on JupyterLite, the dataset needs to be downloaded to the interface using the provided code below.
</p>


The functions below will download the dataset into your browser:


In [3]:
#from pyodide.http import pyfetch

#async def download(url, filename):
#    response = await pyfetch(url)
#    if response.status == 200:
#        with open(filename, "wb") as f:
#            f.write(await response.bytes())

In [4]:
#file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv"

To obtain the dataset, utilize the download() function as defined above:  


In [5]:
#await download(file_path, "m1_survey_data.csv")
file_name="m1_survey_data.csv"

Utilize the Pandas method read_csv() to load the data into a dataframe.


In [6]:
df = pd.read_csv(file_name)

> Note: This version of the lab is working on JupyterLite, which requires the dataset to be downloaded to the interface.While working on the downloaded version of this notebook on their local machines(Jupyter Anaconda), the learners can simply **skip the steps above,** and simply use the URL directly in the `pandas.read_csv()` function. You can uncomment and run the statements in the cell below.


In [7]:
#df = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DA0321EN-SkillsNetwork/LargeData/m1_survey_data.csv")

## Finding duplicates


In this section you will identify duplicate values in the dataset.


 Find how many duplicate rows exist in the dataframe.


In [8]:
# your code goes here

# Show the duplicated rows
print('Duplicated rows:\n')
df_duplicates = df[df.duplicated()].head()
df_duplicates

Duplicated rows:



Respondent                      MainBranch Hobbyist  \
1168        2339  I am a developer by profession      Yes   
1169        2342  I am a developer by profession      Yes   
1170        2343  I am a developer by profession      Yes   
1171        2344  I am a developer by profession      Yes   
1172        2347  I am a developer by profession      Yes   

                                            OpenSourcer  \
1168                         Once a month or more often   
1169                                              Never   
1170  Less than once a month but more than once per ...   
1171                                              Never   
1172                                              Never   

                                             OpenSource          Employment  \
1168  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
1169  The quality of OSS and closed source software ...  Employed full-time   
1170  OSS is, on average, of LOWER quality than prop...  Employed full-time   
1171  The quality of OSS and closed source software ...  Employed full-time   
1172  OSS is, on average, of HIGHER quality than pro...  Employed full-time   

             Country Student  \
1168   United States      No   
1169  United Kingdom      No   
1170          Canada      No   
1171   United States      No   
1172  United Kingdom      No   

                                                EdLevel  \
1168  Some college/university study without earning ...   
1169  Some college/university study without earning ...   
1170        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
1171           Bachelor’s degree (BA, BS, B.Eng., etc.)   
1172        Master’s degree (MA, MS, M.Eng., MBA, etc.)   

                                         UndergradMajor  ...  \
1168  Computer science, computer engineering, or sof...  ...   
1169  Information systems, information technology, o...  ...   
1170  Computer science, computer engineering, or sof...  ...   
1171  Computer science, computer engineering, or sof...  ...   
1172  Computer science, computer engineering, or sof...  ...   

                                 WelcomeChange  \
1168   Just as welcome now as I felt last year   
1169  Somewhat more welcome now than last year   
1170  Somewhat more welcome now than last year   
1171   Just as welcome now as I felt last year   
1172   Just as welcome now as I felt last year   

                                           SONewContent   Age Gender Trans  \
1168                                                NaN  24.0    Man    No   
1169  Tech meetups or events in your area;Courses on...  24.0    Man    No   
1170  Tech articles written by other developers;Indu...  27.0    Man    No   
1171  Tech articles written by other developers;Indu...  24.0    Man    No   
1172                                                NaN   NaN  Woman    No   

                    Sexuality  \
1168  Straight / Heterosexual   
1169  Straight / Heterosexual   
1170  Straight / Heterosexual   
1171  Straight / Heterosexual   
1172  Straight / Heterosexual   

                                              Ethnicity Dependents  \
1168                       White or of European descent         No   
1169                       White or of European descent         No   
1170  Black or of African descent;White or of Europe...         No   
1171                       White or of European descent         No   
1172                                           Biracial         No   

               SurveyLength                  SurveyEase  
1168  Appropriate in length                        Easy  
1169               Too long                        Easy  
1170  Appropriate in length  Neither easy nor difficult  
1171  Appropriate in length                        Easy  
1172               Too long                        Easy  

[5 rows x 85 columns]

In [9]:
# your code goes here

# Find how many duplicate rows exist in the dataframe.
print('Number of duplicated rows:', df.duplicated().sum())

Number of duplicated rows: 154


In [10]:
# Data shape for the original df
df.shape

(11552, 85)

## Removing duplicates


Remove the duplicate rows from the dataframe.


In [11]:
# your code goes here
# Remove duplicated rows from the df
df.drop_duplicates(inplace=True)

Verify if duplicates were actually dropped.


In [12]:
# Data shape for the df after removing duplicated values
df.shape

(11398, 85)

In [13]:
# your code goes here
print('Number of duplicated rows for the new df:', df.duplicated().sum())

Number of duplicated rows for the new df: 0


We have removed 154 duplicated rows from the data frame.

## Finding Missing values


Find the missing values for all columns.


In [14]:
# your code goes here
# Find the total number of missing values for each column 
df.isnull().sum()

Respondent        0
MainBranch        0
Hobbyist          0
OpenSourcer       0
OpenSource       81
               ... 
Sexuality       542
Ethnicity       675
Dependents      140
SurveyLength     19
SurveyEase       14
Length: 85, dtype: int64

Find out how many rows are missing in the column 'WorkLoc'


In [15]:
# your code goes here
df['WorkLoc'].isnull().sum()

32

### **Run code for the test checklist**

In [23]:
# How many rows are there under column 'EdLevel' after removing duplicates?
df['EdLevel'].isnull().sum()

112

In [24]:
# How many rows are there under column 'Country' after removing duplicates?
df['Country'].isnull().sum()

0

In [21]:
# What is the majority category under the column Employment
df['Employment'].value_counts().idxmax()

'Employed full-time'

In [22]:
# Under the column " UndergradMajor", which category has the minimum number of rows?
df['UndergradMajor'].value_counts().idxmin()

'A health science (ex. nursing, pharmacy, radiology)'

In [34]:
# The column ‘ConvertedComp’ contains the annual compensation of the survey respondents. What is the best approach to impute the missing values in this column? 
print(df['ConvertedComp'].isnull().sum(),'\n')
print(df['ConvertedComp'].value_counts(),'\n')
print(df['ConvertedComp'].value_counts().idxmax(),'\n')

816 

ConvertedComp
2000000.0    138
1000000.0    105
100000.0      99
150000.0      92
120000.0      86
            ... 
79183.0        1
20475.0        1
288732.0       1
22056.0        1
19880.0        1
Name: count, Length: 3515, dtype: int64 

2000000.0 



The compensation values are highly skewed, with a few very high values and many lower values.
<br> =>  Since the data is skewed, the median is less affected by outliers compared to the mean.

In [38]:
# How many unique values are there in the CompFreq column?
df['CompFreq'].nunique()

3

In [42]:
# After removing the duplicate rows, how many respondents are being paid yearly?
# Count respondents paid yearly
df[df['CompFreq'] == 'Yearly'].shape[0]

6073

## Imputing missing values


Find the  value counts for the column WorkLoc.


In [33]:
import numpy as np

In [34]:
# your code goes here
df["WorkLoc"].value_counts()

WorkLoc
Office                                            6806
Home                                              3589
Other place, such as a coworking space or cafe     971
Name: count, dtype: int64

Identify the value that is most frequent (majority) in the WorkLoc column.


In [35]:
#make a note of the majority value here, for future reference

# Find the most frequent Work Locatoin value
WorkLoc_mode = df['WorkLoc'].value_counts().idxmax()

Impute (replace) all the empty rows in the column WorkLoc with the value that you have identified as majority.


In [39]:
# your code goes here
df.replace({'WorkLoc': WorkLoc_mode},inplace=True)
# The same with: df['WorkLoc'].replace(np.nan, WorkLoc_mode, inplace=True)
## OR: df['WorkLoc'].fillna(WorkLoc_mode, inplace=True)

After imputation there should ideally not be any empty rows in the WorkLoc column.


Verify if imputing was successful.


In [41]:
# your code goes here
df['WorkLoc'].isnull().sum()

0

## Normalizing data


There are two columns in the dataset that talk about compensation.

One is "CompFreq". This column shows how often a developer is paid (Yearly, Monthly, Weekly).

The other is "CompTotal". This column talks about how much the developer is paid per Year, Month, or Week depending upon his/her "CompFreq". 

This makes it difficult to compare the total compensation of the developers.

In this section you will create a new column called 'NormalizedAnnualCompensation' which contains the 'Annual Compensation' irrespective of the 'CompFreq'.

Once this column is ready, it makes comparison of salaries easy.


<hr>


List out the various categories in the column 'CompFreq'


In [43]:
# your code goes here
print(df['CompFreq'].unique())
print(df['CompFreq'].value_counts())

['Yearly' 'Monthly' 'Weekly' nan]
CompFreq
Yearly     6073
Monthly    4788
Weekly      331
Name: count, dtype: int64


Create a new column named 'NormalizedAnnualCompensation'. Use the hint given below if needed.


Double click to see the **Hint**.

<!--

Use the below logic to arrive at the values for the column NormalizedAnnualCompensation.

If the CompFreq is Yearly then use the exising value in CompTotal
If the CompFreq is Monthly then multiply the value in CompTotal with 12 (months in an year)
If the CompFreq is Weekly then multiply the value in CompTotal with 52 (weeks in an year)

-->


In [48]:
# your code goes here
# Logic:
#1. If the CompFreq is Yearly then use the exising value in CompTotal
#2. If the CompFreq is Monthly then multiply the value in CompTotal with 12 (months in an year)
#3. If the CompFreq is Weekly then multiply the value in CompTotal with 52 (weeks in an year)

# Function to normalize compensation based on frequency
def normalize_compensation(row):
    if row['CompFreq'] == 'Yearly':
        return row['CompTotal']
    elif row['CompFreq'] == 'Monthly':
        return row['CompTotal'] * 12
    elif row['CompFreq'] == 'Weekly':
        return row['CompTotal'] * 52
    else:
        return None  # Handle missing or invalid values

# Apply the function to create the new column
df['NormalizedAnnualCompensation'] = df.apply(normalize_compensation, axis=1) # This applies the normalize_compensation function to each row of the DataFrame. 
                                                                            ## The axis=1 argument specifies that the function should be applied row-wise.

df.head()


Respondent                      MainBranch Hobbyist  \
0           4  I am a developer by profession       No   
1           9  I am a developer by profession      Yes   
2          13  I am a developer by profession      Yes   
3          16  I am a developer by profession      Yes   
4          17  I am a developer by profession      Yes   

                                         OpenSourcer  \
0                                              Never   
1                         Once a month or more often   
2  Less than once a month but more than once per ...   
3                                              Never   
4  Less than once a month but more than once per ...   

                                          OpenSource          Employment  \
0  The quality of OSS and closed source software ...  Employed full-time   
1  The quality of OSS and closed source software ...  Employed full-time   
2  OSS is, on average, of HIGHER quality than pro...  Employed full-time   
3  The quality of OSS and closed source software ...  Employed full-time   
4  The quality of OSS and closed source software ...  Employed full-time   

          Country Student                                            EdLevel  \
0   United States      No           Bachelor’s degree (BA, BS, B.Eng., etc.)   
1     New Zealand      No  Some college/university study without earning ...   
2   United States      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
3  United Kingdom      No        Master’s degree (MA, MS, M.Eng., MBA, etc.)   
4       Australia      No           Bachelor’s degree (BA, BS, B.Eng., etc.)   

                                      UndergradMajor  ...  \
0  Computer science, computer engineering, or sof...  ...   
1  Computer science, computer engineering, or sof...  ...   
2  Computer science, computer engineering, or sof...  ...   
3                                                NaN  ...   
4  Computer science, computer engineering, or sof...  ...   

                                        SONewContent   Age Gender Trans  \
0  Tech articles written by other developers;Indu...  22.0    Man    No   
1                                                NaN  23.0    Man    No   
2  Tech articles written by other developers;Cour...  28.0    Man    No   
3  Tech articles written by other developers;Indu...  26.0    Man    No   
4  Tech articles written by other developers;Indu...  29.0    Man    No   

                 Sexuality                              Ethnicity Dependents  \
0  Straight / Heterosexual           White or of European descent         No   
1                 Bisexual           White or of European descent         No   
2  Straight / Heterosexual           White or of European descent        Yes   
3  Straight / Heterosexual           White or of European descent         No   
4  Straight / Heterosexual  Hispanic or Latino/Latina;Multiracial         No   

            SurveyLength                  SurveyEase  \
0  Appropriate in length                        Easy   
1  Appropriate in length  Neither easy nor difficult   
2  Appropriate in length                        Easy   
3  Appropriate in length  Neither easy nor difficult   
4  Appropriate in length                        Easy   

  NormalizedAnnualCompensation  
0                      61000.0  
1                     138000.0  
2                      90000.0  
3                     348000.0  
4                      90000.0  

[5 rows x 86 columns]

In [49]:
# Find the median of 'NormalizedAnnualCompensation'
df['NormalizedAnnualCompensation'].median()

100000.0

## Authors


Ramesh Sannareddy


### Other Contributors


Rav Ahuja


 ## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2024-09-24|1.1|Madhusudhan Moole|Updated lab|
|2024-09-23|1.0|Raghul Ramesh|Created lab|


<!--| Date (YYYY-MM-DD) | Version | Changed By        | Change Description                 |
| ----------------- | ------- | ----------------- | ---------------------------------- |
| 2020-10-17        | 0.1     | Ramesh Sannareddy | Created initial version of the lab |--!>


## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
